src/cmd-aws-replicate

```diff
@@ -68,19 +68,30 @@ def run_ore():
     if args.log_level:
         ore_args.extend(['--log-level', args.log_level])
     ore_args.extend(['aws', 'copy-image', '--image', source_image, '--region', source_region])
-    ore_args.extend(region_list)
-    print("+ {}".format(subprocess.list2cmdline(ore_args)))
-    ore_data = json.loads(subprocess.check_output(ore_args))
-    # This matches the Container Linux schema:
-    # https://stable.release.core-os.net/amd64-usr/current/coreos_production_ami_all.json
-    ami_data = [{'name': region,
-                 'hvm': vals['ami'],
-                 'snapshot': vals['snapshot']}
-                for region, vals in ore_data.items()]
-    buildmeta['amis'].extend(ami_data)
-    write_json(buildmeta_path, buildmeta)
+
+    upload_failed_in_regions = []
+    for upload_region in region_list:
+        region_ore_args = ore_args.copy().append(upload_region)
+        print("+ {}".format(subprocess.list2cmdline(region_ore_args)))
+        try:
+            ore_data = json.loads(subprocess.check_output(region_ore_args))
+        except subprocess.CalledProcessError:
+            upload_failed_in_regions.append(upload_region)
+            continue
```

Comment:
I'm not sure about that - it has already been uploaded to the first region. If the build is considered failed we should be removing existing AMI and key from the bucket

We could attempt to do a retry here instead of simply skipping it later on

Reply:
We don't have a good story on cleanup right now; we know we need to add something like `cosa build-remove` or so, but let's not try to scope it into this PR.

From the current RHCOS pipeline perspective, if a region isn't uploaded, the build is failed because it'd mean `openshift-install` wouldn't work.

As far as retrying...dunno.  We already have a retry loop in the higher level Jenkins - we'll do another build later.  We're not consistent on this, but I in general lean towards making the higher level retry (Jenkins jobs) work better rather than scattering retries inside individual builds too much.

Here's a good example of why: It doesn't make sense to e.g. spend 24 hours retrying an AMI upload for our development pipeline, because by that time we almost certainly have *new* content we want to upload anyways.

Now there is the flip side of we have a "final" build we want to get out and it's just failing in one region - that's where we may want to retry.  In the end though our code/rpms for releases gets pretty frozen so the next build even though it might gain a new version number would be pretty much the same.

Reply:
Okay, make sense to me. The build would fail anyway, but it would stop when AMI cannot be uploaded to a region - and meta.json would keep the list of previously uploaded AMIs (so that these could be GC'ed later)

Reply:
> If a region failed, I think that build is considered broken.

Not necessarily IMO, it can also fail for legitimate reasons that are not related to the build itself, e.x.: the cloud provider is throwing 500s in that region temporarily.

> As far as retrying...dunno. We already have a retry loop in the higher level Jenkins - we'll do another build later. We're not consistent on this, but I in general lean towards making the higher level retry (Jenkins jobs) work better rather than scattering retries inside individual builds too much.

I do agree that retries in jenkins are preferred (and that code should be able to handle being re-ran). For this in particular I'd lean more towards having job level retries rather than immediately failing the build entirely and requiring a new build.

> Now there is the flip side of we have a "final" build we want to get out and it's just failing in one region - that's where we may want to retry. In the end though our code/rpms for releases gets pretty frozen so the next build even though it might gain a new version number would be pretty much the same.

Yeah, I think a lot of this will end up coming down to how we design the build pipeline itself. As I've mentioned in other forums I'd be really happy to see the pipeline moving into a less monolithic set of tasks that are individually retry-able (a la the CL `jenkins-os` setup) and don't affect the greater build status, especially for things involving external cloud providers that can (and do) fail for reasons beyond our control.

Reply:
> Not necessarily IMO, it can also fail for legitimate reasons that are not related to the build itself, e.x.: the cloud provider is throwing 500s in that region temporarily.

The point is though that for RHCOS we couldn't use that build for OpenShift, so it failed.  AFAIK the idea for FCOS is to require image uploads too as much as possible.

Reply:
> The point is though that for RHCOS we couldn't use that build for OpenShift, so it failed. AFAIK the idea for FCOS is to require image uploads too as much as possible.

I'm not sure I understand why you wouldn't be able to, assuming that the images do end up being fully replicated (via retries) before sending it over to OpenShift.

Reply:
> I'm not sure I understand why you wouldn't be able to, assuming that the images do end up being fully replicated (via retries) before sending it over to OpenShift.

I think it's simpler to consider that build failed and make a new one.